ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java

```diff
@@ -33,5 +42,138 @@ static boolean haveEquivalentFields(BlobProperties a, BlobProperties b) {
+   *                                there are more values in this array than there are servers, an exception is thrown.
+   *                                If there are less, the rest of the servers are set to
+   *                                {@link ServerErrorCode#No_Error}
+   * @param serverLayout A {@link MockServerLayout} that is used to find the {@link MockServer}s to set error codes on.
+   * @param expectedError The {@link RouterErrorCode} expected, or null if no error is expected.
+   * @param errorCodeChecker Performs the checks that ensure that the expected {@link RouterErrorCode} is returned .
+   * @throws Exception
+   */
+  static void testWithErrorCodes(ServerErrorCode[] serverErrorCodesInOrder, MockServerLayout serverLayout,
+      RouterErrorCode expectedError, ErrorCodeChecker errorCodeChecker)
+      throws Exception {
+    setServerErrorCodes(serverErrorCodesInOrder, serverLayout);
+    errorCodeChecker.check(expectedError);
+    resetServerErrorCodes(serverLayout);
+  }
+
+  /**
+   * Test that an operation returns a certain result when each server in a partition returns a certain error code.
+   * @param serverErrorCodesInOrder The error codes to set in the order of the servers returned by
+   *                                {@link PartitionId#getReplicaIds()}. If there are more values in this array than
+   *                                there are servers, an exception is thrown. If there are less, the rest of the
+   *                                servers are set to {@link ServerErrorCode#No_Error}
+   * @param serverLayout A {@link MockServerLayout} that is used to find the {@link MockServer}s to set error codes on.
+   * @param expectedError The {@link RouterErrorCode} expected, or null if no error is expected.
+   * @param errorCodeChecker Performs the checks that ensure that the expected {@link RouterErrorCode} is returned .
+   * @throws Exception
+   */
+  static void testWithErrorCodes(ServerErrorCode[] serverErrorCodesInOrder, PartitionId partition,
+      MockServerLayout serverLayout, RouterErrorCode expectedError, ErrorCodeChecker errorCodeChecker)
+      throws Exception {
+    setServerErrorCodes(serverErrorCodesInOrder, partition, serverLayout);
+    errorCodeChecker.check(expectedError);
+    resetServerErrorCodes(serverLayout);
+  }
+
+  /**
+   * Set the servers in the specified layout to respond with the designated error codes.
+   * @param serverErrorCodesInOrder The error codes to set in the order of the servers in {@code serverLayout}.  If
+   *                                there are more values in this array than there are servers, an exception is thrown.
+   *                                If there are less, the rest of the servers are set to
+   *                                {@link ServerErrorCode#No_Error}
+   * @param serverLayout A {@link MockServerLayout} that is used to find the {@link MockServer}s to set error codes on.
+   * @throws Exception
```

Comment:
does not. Maybe make it `@throws IllegalArgumentException` and describe when that happens ? (instead of doing it in the main doc).